File: app/src/main/java/com/boardgamegeek/ui/widget/GameRankRow.java
package com.boardgamegeek.ui.widget;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.view.LayoutInflater;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.boardgamegeek.R;
import com.boardgamegeek.util.ColorUtils;
import com.boardgamegeek.util.PresentationUtils;

import butterknife.BindView;
import butterknife.ButterKnife;

public class GameRankRow extends LinearLayout {
	@BindView(R.id.rank) TextView rankView;
	@BindView(R.id.name) TextView nameView;
	@BindView(R.id.rating) TextView ratingView;

	public GameRankRow(Context context, boolean isFamily) {
		super(context);
		@LayoutRes int layout = isFamily ? R.layout.row_game_rank_family : R.layout.row_game_rank_subtype;
		LayoutInflater.from(context).inflate(layout, this);
		ButterKnife.bind(this);
	}

	public void setRank(int rank) {
		if (PresentationUtils.isRankValid(rank)) {
			rankView.setText(getContext().getString(R.string.rank_prefix, rank));
			rankView.setVisibility(VISIBLE);
		} else {
			rankView.setVisibility(INVISIBLE);
		}
	}

	public void setName(CharSequence name) {
		nameView.setText(name);
	}

	public void setRatingView(double rating) {
		ratingView.setText(PresentationUtils.describeAverageRating(getContext(), rating));
		ColorUtils.setTextViewBackground(ratingView, ColorUtils.getRatingColor(rating));
	}
}

File: app/src/main/java/com/boardgamegeek/ui/widget/GameRankRow.kt
package com.boardgamegeek.ui.widget

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import com.boardgamegeek.R
import com.boardgamegeek.databinding.RowGameRankSubtypeBinding
import com.boardgamegeek.extensions.*
import java.text.DecimalFormat

@SuppressLint("ViewConstructor")
class GameRankRow(context: Context, isFamily: Boolean) : LinearLayout(context) {
    private val binding = RowGameRankSubtypeBinding.inflate(LayoutInflater.from(context), this, true)

    init {
        if (isFamily) {
            @Suppress("DEPRECATION")
            binding.rankView.setTextAppearance(context, R.style.Text)
            @Suppress("DEPRECATION")
            binding.nameView.setTextAppearance(context, R.style.Text)
            binding.ratingView.setTypeface(binding.ratingView.typeface, Typeface.NORMAL)
        } else {
            @Suppress("DEPRECATION")
            binding.rankView.setTextAppearance(context, R.style.Text_Subtitle)
            @Suppress("DEPRECATION")
            binding.nameView.setTextAppearance(context, R.style.Text_Subtitle)
            binding.ratingView.setTypeface(binding.ratingView.typeface, Typeface.BOLD)
        }
    }

    fun setRank(rank: Int) {
        if (rank.isRankValid()) {
            binding.rankView.text = context.getString(R.string.rank_prefix, rank)
            binding.rankView.visibility = View.VISIBLE
        } else {
            binding.rankView.visibility = View.INVISIBLE
        }
    }

    fun setName(name: CharSequence) {
        binding.nameView.text = name
    }

    fun setRatingView(rating: Double) {
        binding.ratingView.text = rating.asScore(context, format = AVERAGE_RATING_FORMAT)
        binding.ratingView.setTextViewBackground(rating.toColor(ratingColors))
    }

    companion object {
        private val AVERAGE_RATING_FORMAT = DecimalFormat("#0.000")
    }
}
